homeassistant/components/braviatv/coordinator.py

```diff
@@ -192,27 +183,68 @@ async def async_update_playing(self) -> None:
         self.media_title = playing_info.get("title")
         self.media_uri = playing_info.get("uri")
         self.media_duration = playing_info.get("durationSec")
-        if program_title := playing_info.get("programTitle"):
-            self.media_title = f"{self.media_title}: {program_title}"
+        self.media_channel = None
+        self.media_content_id = None
+        self.media_content_type = None
+        self.source = None
         if self.media_uri:
-            source = self.source_map.get(self.media_uri, {})
-            self.source = source.get("title")
-            self.is_channel = self.media_uri[:2] == "tv"
-            if self.is_channel:
+            self.media_content_id = self.media_uri
+            if self.media_uri[:8] == "extInput":
+                self.source = playing_info.get("title")
+            if self.media_uri[:2] == "tv":
+                self.media_title = playing_info.get("programTitle")
+                self.media_channel = playing_info.get("title")
                 self.media_content_id = playing_info.get("dispNum")
                 self.media_content_type = MediaType.CHANNEL
-            else:
-                self.media_content_id = self.media_uri
-                self.media_content_type = None
-        else:
-            self.source = None
-            self.is_channel = False
-            self.media_content_id = None
-            self.media_content_type = None
         if not playing_info:
             self.media_title = "Smart TV"
             self.media_content_type = MediaType.APP
 
+    async def async_update_sources(self) -> None:
+        """Update all sources."""
+        self.source_list = []
+        self.source_map = {}
+
+        inputs = await self.client.get_external_status()
+        self._sources_extend(inputs, SourceType.INPUT, add_to_list=True)
+
+        apps = await self.client.get_app_list()
+        self._sources_extend(apps, SourceType.APP, sort_by="title")
+
+        channels = await self.client.get_content_list_all("tv")
+        self._sources_extend(channels, SourceType.CHANNEL)
```

Comment:
What error handling is in place if the updating gets wrong? Should there be a `try` block here?

Reply:
never mind. I see _async_update_data is guarding exceptions